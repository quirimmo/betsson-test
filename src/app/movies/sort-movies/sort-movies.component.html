<form>
	<div class="row sort-movies-container">
		<mat-form-field
			class="col-lg-3 offset-lg-9 col-md-4 offset-md-8 col-sm-5 offset-sm-7 col-12 w-100"
		>
			<mat-select
				[formControl]="sortBySelect"
				placeholder="Sort by"
				name="movie-sort-by"
			>
				<mat-option class="sort-movies-options" *ngFor="let item of orderingOptions" [value]="item">
					{{ item.label }}
				</mat-option>
			</mat-select>
		</mat-form-field>
	</div>
</form>
