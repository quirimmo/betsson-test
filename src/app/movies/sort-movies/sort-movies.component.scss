.sort-movies-container {
  margin: 0px;
  text-align: center;
}
