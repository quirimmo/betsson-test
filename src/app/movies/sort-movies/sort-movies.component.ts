import { Component, OnInit } from '@angular/core';
import { FormControl } from '@angular/forms';
import { SortingValues } from '../pipes/sort-movies.pipe';

interface MovieSortingValues extends SortingValues {
	label: string;
}

@Component({
	selector: 'sort-movies',
	inputs: [],
	styleUrls: ['./sort-movies.component.scss'],
	templateUrl: './sort-movies.component.html'
})
export class SortMoviesComponent implements OnInit {
	sortBySelect = new FormControl();

	orderingOptions: MovieSortingValues[] = [
		{ value: 'id', isAscending: true, label: 'Default (Ascending)' },
		{ value: 'id', isAscending: false, label: 'Default (Descending)' },
		{ value: 'name', isAscending: true, label: 'Name (Ascending)' },
		{ value: 'name', isAscending: false, label: 'Name (Descending)' },
		{ value: 'rate', isAscending: true, label: 'Rate (Ascending)' },
		{ value: 'rate', isAscending: false, label: 'Rate (Descending)' },
		{ value: 'length', isAscending: true, label: 'Duration (Ascending)' },
		{ value: 'length', isAscending: false, label: 'Duration (Descending)' }
	];

	constructor() {}

	ngOnInit(): void {
		this.sortBySelect.setValue(this.orderingOptions[0]);
	}
}
