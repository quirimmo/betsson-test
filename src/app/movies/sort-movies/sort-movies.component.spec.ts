import { TestBed, async } from '@angular/core/testing';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { SortMoviesComponent } from './sort-movies.component';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { MatFormFieldModule, MatSelectModule } from '@angular/material';
import { MoviesRouterService } from '../movies-router.service';

const mockGetSortByParam = jest
	.fn()
	.mockReturnValue({ sortBy: 'name', isAscending: false });
const mockAppendSortByParam = jest.fn();
const mockMoviesRouterService = {
	provide: MoviesRouterService,
	useValue: {
		getSortByParam: mockGetSortByParam,
		appendSortByParam: mockAppendSortByParam
	}
};
const providers = [mockMoviesRouterService];
const imports = [
	BrowserAnimationsModule,
	FormsModule,
	ReactiveFormsModule,
	MatSelectModule,
	MatFormFieldModule
];
const declarations = [SortMoviesComponent];

describe('SortMovies Component', () => {
	let componentFixture, componentInstance;

	beforeEach(async(() => {
		TestBed.configureTestingModule({
			declarations,
			providers,
			imports
		}).compileComponents();
		componentFixture = TestBed.createComponent(SortMoviesComponent);
		componentInstance = componentFixture.debugElement.componentInstance;
	}));
	afterEach(() => {
		jest.clearAllMocks();
		jest.restoreAllMocks();
	});

	it('should create the component', () => {
		expect(componentFixture).toBeTruthy();
	});

	it('should match the snapshot', () => {
		expect(componentFixture).toMatchSnapshot();
	});

	it('should define the orderingOptions', () => {
		expect(componentInstance.orderingOptions).toEqual([
			{ value: 'id', isAscending: true, label: 'Default (Ascending)' },
			{ value: 'id', isAscending: false, label: 'Default (Descending)' },
			{ value: 'name', isAscending: true, label: 'Name (Ascending)' },
			{ value: 'name', isAscending: false, label: 'Name (Descending)' },
			{ value: 'rate', isAscending: true, label: 'Rate (Ascending)' },
			{ value: 'rate', isAscending: false, label: 'Rate (Descending)' },
			{ value: 'length', isAscending: true, label: 'Duration (Ascending)' },
			{ value: 'length', isAscending: false, label: 'Duration (Descending)' }
		]);
	});

	describe('OnInit', () => {
		beforeEach(async () => {
			componentFixture.detectChanges();
			await componentFixture.whenStable();
		});

		it('should call the getSortByParam method of moviesRouterService', () => {
			expect(mockGetSortByParam).toHaveBeenCalled();
		});

		it('should init the sortBySelect value', () => {
			expect(componentInstance.sortBySelect.value).toEqual({
				value: 'name',
				isAscending: false,
				label: 'Name (Descending)'
			});
		});

		it('should init the sortBySelect value to the default ordering option', async () => {
			mockGetSortByParam.mockReturnValue({ sortBy: null, isAscending: null });
			componentInstance.ngOnInit();
			componentFixture.detectChanges();
			await componentFixture.whenStable();
			expect(componentInstance.sortBySelect.value).toEqual({
				value: 'id',
				isAscending: true,
				label: 'Default (Ascending)'
			});
		});
	});

	describe('OnDestroy', () => {
		it('should unsubscribe to sortBySelectChangesSubscriber', () => {
			const spy = jest.fn();
			componentInstance.sortBySelectChangesSubscriber = {
				unsubscribe: spy
			};
			componentInstance.ngOnDestroy();
			expect(spy).toHaveBeenCalled();
		});
	});

	describe('onChangeSortBy', () => {
		it('should call the mockAppendSortByParam', () => {
			componentInstance.sortBySelect.value = {
				value: 'length',
				isAscending: true,
				label: 'Duration (Ascending)'
			};
			componentInstance.onChangeSortBy();
			expect(mockAppendSortByParam).toHaveBeenCalledWith('length', true);
		});

		it('should call the onChangeSortBy method if the value of sortBySelect changes', async () => {
			const spy = jest.spyOn(componentInstance, 'onChangeSortBy');
			componentInstance.sortBySelect.setValue({
				value: 'length',
				isAscending: true,
				label: 'Duration (Ascending)'
			});
			componentFixture.detectChanges();
			await componentFixture.whenStable();
			expect(spy).toHaveBeenCalled();
		});
	});
});
