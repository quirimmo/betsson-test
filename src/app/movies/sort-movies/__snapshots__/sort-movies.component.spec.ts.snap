// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortMovies Component should match the snapshot 1`] = `
<sort-movies
  moviesRouterService={[Function Object]}
  orderingOptions={[Function Array]}
  sortBySelect={[Function FormControl]}
>
  <form
    novalidate=""
  >
    <div
      class="row sort-movies-container"
    >
      <mat-form-field
        class="col-lg-3 offset-lg-9 col-md-4 offset-md-8 col-sm-5 offset-sm-7 col-12 w-100 mat-form-field ng-tns-c3-0 mat-primary"
      >
        <div
          class="mat-form-field-wrapper"
        >
          <div
            class="mat-form-field-flex"
          >
            
            
            <div
              class="mat-form-field-infix"
            >
              <mat-select
                class="mat-select ng-tns-c4-1"
                name="movie-sort-by"
                placeholder="Sort by"
                role="listbox"
              >
                <div
                  aria-hidden="true"
                  cdk-overlay-origin=""
                  class="mat-select-trigger"
                >
                  <div
                    class="mat-select-value"
                  >
                    
                    
                  </div>
                  <div
                    class="mat-select-arrow-wrapper"
                  >
                    <div
                      class="mat-select-arrow"
                    />
                  </div>
                </div>
                
              </mat-select>
              <span
                class="mat-form-field-label-wrapper"
              >
                
              </span>
            </div>
            
          </div>
          
          <div
            class="mat-form-field-subscript-wrapper"
          >
            
            
          </div>
        </div>
      </mat-form-field>
    </div>
  </form>
</sort-movies>
`;
